### DISCUSSION TITLE
Hit Postgres 'Written Data' limit without writing data

### DISCUSSION THREAD

**Discussion by InternetJohnny** [vercel/community#4815] [Help] [answered]
### Summary

I've started developing a project and decided to use Vercel Postgres.

All I did was work on my Prisma Schema, and I performed 5 migrations, that's all.
My schema is small, and only has 8 models.
I have not written or read any database records.

Yet on my Vercel dashboard, it shows that I've hit my Written Data cap, being at 353 MB / 256 MB.

How could this happen?

### Steps to Reproduce

- Create Postgres database in Vercel
- Connect to project
- Create a Prisma schema with 8 models
- Perform 5 migrations with minor changes

**Comment by Maxim-Mazurok** [comment]
Same happened to me twice already as well

**Comment by amyegan** [comment]
Data added, data updated, data deleted and migrations can all impact written data usage. I don't have enough info here to determine the exact cause of high written data usage. 

With your permission the engineers can get read-only access to investigate further. If you are okay with that, please let me know the name/ID of the database we should investigate 🙏

**Reply by InternetJohnny** [reply]
Hi Amy. I have seen this reported quite a bit, don't believe I'm the only one with this issue. Were you unable to replicate it with the instructions provided?

**Reply by amyegan** [reply]
I am not able to replicate the problem you described. 

I know there was a [data reporting problem back in June](https://github.com/orgs/vercel/discussions/2981) that was solved. And there have been a few reports of confusing storage numbers that were explained by actual usage. But I'm sure about other unanswered reports. If you can share links, I would like to investigate them as well. 🙏

**Reply by Maxim-Mazurok** [reply]
@amyegan feel free to investigate mine: `ep-late-mode-33358376-pooler.us-east-1.postgres.vercel-storage.com` under `maxim@mazurok.com` account.
It has three tiny tables and negligible activity, yet sitting at 201 MB / 256 MB Written Data as of now.
Hope this helps, cheers!

**Reply by bmmathe** [reply]
@amyegan you can look at mine too.  We are locked out of database even though we have < 100 records and we haven't even really been using it (we use a separate self-hosted dev database).
ep-flat-snowflake-40387691-pooler.us-east-1.postgres.vercel-storage.com

**Reply by amyegan** [reply, marked as answer]
Thank you all for waiting. I checked with our Storage engineering team to get more information about what is happening here.

Keep in mind that written data is not only data added. Data updated, data deleted, and migrations can also generate written data. So running multiple migrations can have a noticeable impact on written data. According to Prisma's [shadow database](https://www.prisma.io/docs/orm/prisma-migrate/understanding-prisma-migrate/shadow-database) documentation, 
> The shadow database is a second, _temporary_ database that is **created and deleted automatically*** each time you run `prisma migrate dev` and is primarily used to **detect problems** such as schema drift or potential data loss of the generated migration.

Based on the information available, this accounts for the high written data usage.

**Reply by Maxim-Mazurok** [reply]
I see, @amyegan , thanks for checking, I was kinda expecting something like that...
While it does make sense technically, it still doesn't make sense from a product usage perspective. You know that most people will probably use prisma in vercel. So it probably makes sense to somehow make this integration smoother. For example, you could check the behaviour, if the database is deleted within a few minutes and has a name/pattern of prisma's shadow database - then it could be not counted towards the limit. Otherwise you see the problem, I didn't have a ton of migrations, I have two migrations:

```SQL
-- CreateTable
CREATE TABLE "User" (
    "id" SERIAL NOT NULL,
    "email" TEXT NOT NULL,
    "createdAt" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,

    CONSTRAINT "User_pkey" PRIMARY KEY ("id")
);

-- CreateIndex
CREATE UNIQUE INDEX "User_email_key" ON "User"("email");
```

and
```SQL
-- AlterTable
ALTER TABLE "User" ALTER COLUMN "createdAt" SET DATA TYPE TIMESTAMPTZ(3);
```

That's it. And yet I almost hit the resource usage limits. Would you expect people with schemas as simple as this one to upgrade? And I have a few reads/writes per month as well. Definitely not at the stage where I can justify paying for the DB. Anyway, these are some suggestions based on the common sense, I'm not sure if it's technically possible or aligns with the business model, that's up to you, thanks for looking into it!

**Reply by bmmathe** [reply]
The shadow database is only used if you issue the command `prisma migrate dev`.  If you issue `prisma migrate deploy`, the shadow database isn't created.  You can see the documentation here: https://www.prisma.io/docs/orm/prisma-migrate/understanding-prisma-migrate/shadow-database  In my case, I use a self-hosted Postgres database so all my migration testing is done on my db.  I don't even have a `prisma migrate` script committed (only `prisma generate` which only creates the Prisma client).

I'm not 100% sure I have never run `prisma migrate dev` on the Vercel database but if I have it was only one-time to get the initial structure stood up (6 tables with about 60 rows in total).  Even running the command once shouldn't put me over 300MB with such a small database.  Unfortunately I can't even check to see what is in my production _prisma_migrations table because I can't access my database.

**Reply by shakedlokits** [reply]
I encountered the same issue with Prisma, as of now, I'm stuck and moving to a different cloud provider for my DB, which is quite disappointing.
When this is resolved, I suggest that you enable affected users to reverse the impact of this bug.

**Comment by danharper** [comment]
Hit the same issue! I set up a Vercel Postgres DB last night, and within just a few hours (<6) I'd hit the "Written Data" warning.

I only have 2 tables (1 of my own + Prisma's migrations table).
My table has just 4 rows, each with <1kb of JSON (so 4kb total).
The migrations table has just 3 rows.

I think I ran migrations, like... 4 times? 3 migrations, plus one "reset".

Really just a _very_ basic setup, and yet I've used 224mb out of the 256mb "Written Data" allowance, and I'm using 66mb of storage (33mb is Postgres itself, it says)?

I'd put my usage at _waayyyyy_ below even what a basic Postgres tutorial would cover - I'd literally just done one page of the Prisma docs 😅

I'm not sure this is from Prisma's usage of a Shadow Database, either -- that seems to involve manual configuration, and wouldn't work on Vercel's Hobby plan where I'm limited to one DB.

(feel free to access my data if needed to debug)

**Comment by tombyers** [comment]
Same story with me. I have just set up a Vercel Postgres DB, using Prisma - literally just getting started. I have a total of 6 rows (!) of data across 2 tables, and yet have apparently used 222 MB / 256 MB written data allowance. The setup offers such a nice workflow, so it's very disappointing to have to look for alternatives already.

**Comment by kuringer** [comment]
I'm experiencing a similar issue. I have just 2 models and performed only 2 migrations, with almost no data in the database. Previously, I was working with the database using Vercel SQL and everything was fine. However, after adding Prisma, the written data skyrocketed and I've now exceeded the limit. It's puzzling how this could happen with such minimal activity.

**Comment by bmmathe** [comment]
One of our engineers was able to track down our issue.  We had a bug in one of our components that ended up causing an infinite loop in a very specific situation (useEffect calling a function that wasn't a properly configured useCallback function) which caused ~160 records to be pulled every hit to the API.  After running the math, 15,000 calls pulling down our small data set seemed to line up with what our quota is showing.  :(

**Comment by WesleyKapow** [comment]
Hitting this now too. I connected to the db using `psql` and `\l+` to list db sizes, the `verceldb` that I'm using is at `7424 kB` which I think is mostly just boilerplate. I have VERY little data and yet written data is listed as `196MB` and storage is somehow at `94MB`. **This doesn't give me much confidence.**

Reported storage: 
<img width="657" alt="Screenshot 2023-12-20 at 5 45 55 PM" src="https://github.com/vercel/community/assets/1568246/30c7d8a2-96a6-4767-98c4-ef8765025d33">

Row counts (I got like no data):
<img width="744" alt="Screenshot 2023-12-20 at 5 49 57 PM" src="https://github.com/vercel/community/assets/1568246/d5a3f04c-5fb9-4f98-8cc5-82dd46f60b07">

**Reply by WesleyKapow** [reply]
OK, I tried directly against neon.tech and see the same behavior of large "written data". I believe the above comments of this is likely due to prisma's shadow database. Prisma being such a common integration, I suggest an integration doc explaining this issue and encouraging use of a local postgres for `prisma migrate dev`.

**Comment by shakedlokits** [comment]
Here's another one: 
1. I have no active deployments or productions branches.
2. last time I've touched the project was last month.
3. I got a "written data" usage pause.
4. I'm not using a shadow database.
```
    "db:migrate": "prisma migrate dev --create-only --name",
    "db:deploy": "prisma migrate deploy",
```
5. **I AM NOT USING YOUR DB...😅**
(Moved to Supabase since my last comment a month ago)

I think you've got a problem there Houston 👌🏼
I suggest that you either learn the Prisma internals or get rid or the really odd DB KPI you've created, it's killing this feature for you guys..
Either way, please stop blaming the customers. They are not idiots. It just makes your Solutionists/Product seem disrespectful.

<img width="1593" alt="image" src="https://github.com/vercel/community/assets/2076274/38808038-79a6-4aa5-930f-a9be6d475a98">

![image](https://github.com/vercel/community/assets/2076274/6144d8c2-2a9a-43cc-ae57-9506a465cb02)

**Comment by hoggworks** [comment]
Let me add to the chorus of people experiencing this: I've hardly got any data in, though I did a few migrations yesterday, I'd already seemingly been using up a pretty good amount of "Written Data" in a very short period of time. 

Extra frustrating is that all access to the DB is paused, with my "Written Data" at 295/250, and my Data Transfer at only 3.62 MB/256 MB. The interface makes it hard to diagnose the problem, as well, unless I'm missing something, since I can't use the "Browse Data" functionality in the admin to see what might be filling up my quota. VERY frustrating experience. 

If this is just because of migrations -- I was using migrate dev because I hadn't realized the extra weight it took up -- will the Pro usage be okay, or will it also skyrocket? Can I nuke the shadow DB, or is the usage of it counted even though it's auto-deleted?

**Comment by ruchernchong** [comment]
This happened to me too when I ran `prisma migrate dev`. This is such a weird experience.

**Comment by token-ed** [comment]
I think something similar is happening with me as well. But upon reaching the "Written Data" limit (Hobby plan - 256 MB), I did not receive any notification, my application simply started to err with `endpoint is temporary unavailable. check your quotas and/or contact our support.`

I'm wondering if this is on Vercel side? I've ran about 5 migrations through Prisma with `prisma migrate dev`, made a couple of insertions and deletes. I've truncated the database a few times too. How can these operations yield to such big sized logs?
On the "[written data](https://vercel.com/docs/storage/vercel-postgres/usage-and-pricing#written-data)" section of the pricing page it's stated that:
- `Each operation is logged, and it is the size of these logs that contributes to your written data volume.`

Along with a couple of examples on the correlation between an "operation" and how its size contributes to the "Written Data" quota:

```
For Example:

- Updating a single field in a row that is 1KB in size, 1000 times, results in 1MB of written data.
- Deleting a 1KB row from a table also counts as writing 1KB of data.
```

My rows of data are fairly simple and I couldn't possibly reach the limit according to my estimation. But the logs confirm that I indeed did reach it:
![image](https://github.com/vercel/community/assets/149210610/2125200d-9459-452a-bda3-6d00e576b625)

Now, I suppose I could move to the Pro plan, but I wanted to confirm first if this is not an issue? I would like to understand how these logs contribute with so much "Written Data".
